### PR TITLE
Support unordered lists in Terraform.

### PR DESCRIPTION
**Pull request by ndmckinley**
Some things are not ordered on the server side.  This supports a customizediff for lists that are not ordered, and applies that to subnetwork's secondary IP ranges.

-----------------------------------------------------------------
# [all]
## [terraform]
Support unordered list on subnetwork secondary IP range.

### CODE REVIEW

File: templates/terraform/unordered_list_customize_diff.erb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    }
+    if n != nil {
+        new = append(new, n)
+    }
+}
+if len(old) != len(new) {
+    return nil
+}
+
+for o, _ := range old {
```
Comment (danawillow) [inline templates/terraform/unordered_list_customize_diff.erb]: This logic seems wrong to me. First off, o and n are the array indices. Second, this looks like it'll return nil (i.e. keep doing what you're doing) as soon as any two elements aren't equal. What if instead, we loaded everything into a schema.Set and then checked equality on that?

Comment (ndmckinley) [reply templates/terraform/unordered_list_customize_diff.erb]: Weird that the tests passed - yikes, thanks for catching that.

Loading into a schema.Set is a good plan, I'll try that.  I tried with regular go sets, but of course that didn't work since `[]interface{}` isn't hashable.

Comment (ndmckinley) [reply templates/terraform/unordered_list_customize_diff.erb]: That's much better, thank you.

File: templates/terraform/resource_definition/subnetwork.erb [hunk 2]
```diff
@@ -1,3 +1,4 @@
 CustomizeDiff: customdiff.All(
 		customdiff.ForceNewIfChange("ip_cidr_range", isShrinkageIpCidr),
+        resourceComputeSubnetworkCustomizeDiff,
```
Comment (danawillow) [inline templates/terraform/resource_definition/subnetwork.erb]: nit: mind making this line tabs or the previous line spaces?

File: products/compute/terraform.yaml [hunk 3]
```diff
@@ -623,6 +623,8 @@ overrides: !ruby/object:Provider::ResourceOverrides
           function: 'validateGCPName'
       secondaryIpRanges: !ruby/object:Provider::Terraform::PropertyOverride
         name: secondaryIpRange
+        unordered_list: true
+        default_from_api: true
```
Comment (danawillow) [inline products/compute/terraform.yaml]: why this change?

File: templates/terraform/unordered_list_customize_diff.erb [hunk 4]
```diff
@@ -0,0 +1,44 @@
+keys := diff.GetChangedKeysPrefix(<%= go_literal(Google::StringUtils.underscore(prop.name)) -%>)
+if len(keys) < 1 {
+	return nil
+}
+
+oldCount, newCount := diff.GetChange("<%= Google::StringUtils.underscore(prop.name) -%>.#")
+var count int
+if oldCount.(int) < newCount.(int) {
```
Comment (danawillow) [inline templates/terraform/unordered_list_customize_diff.erb]: Does it make sense to return here if the counts aren't the same?

File: templates/terraform/unordered_list_customize_diff.erb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    count = newCount.(int)
+} else {
+    count = oldCount.(int)
+}
+if count < 1 {
+    return nil
+}
+
+if count < 1 {
```
Comment (danawillow) [inline templates/terraform/unordered_list_customize_diff.erb]: This is the same as the lines immediately before it

File: templates/terraform/unordered_list_customize_diff.erb (hunk 4 above)
Comment (ndmckinley) [reply templates/terraform/unordered_list_customize_diff.erb]: Duplicates may exist.  Added a comment.

File: products/compute/terraform.yaml (hunk 3 above)
Comment (ndmckinley) [reply products/compute/terraform.yaml]: The problem comes from the fact that we're implicitly comparing against the returned value from the API - we wouldn't have any diff problems otherwise.

File: templates/terraform/resource_definition/subnetwork.erb (hunk 2 above)
Comment (ndmckinley) [reply templates/terraform/resource_definition/subnetwork.erb]: Done.

File: templates/terraform/unordered_list_customize_diff.erb (hunk 5 above)
Comment (ndmckinley) [reply templates/terraform/unordered_list_customize_diff.erb]: Fixed!

File: templates/terraform/unordered_list_customize_diff.erb [hunk 6]
```diff
@@ -0,0 +1,37 @@
+keys := diff.GetChangedKeysPrefix(<%= go_literal(Google::StringUtils.underscore(prop.name)) -%>)
+if len(keys) > 0 {
```
Comment (danawillow) [inline templates/terraform/unordered_list_customize_diff.erb]: nit: I would personally find it easier to follow if it were
```
if len(keys) == 0 {
  return nil
}
```
but up to you

File: templates/terraform/unordered_list_customize_diff.erb [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    oldCount, newCount := diff.GetChange("<%= Google::StringUtils.underscore(prop.name) -%>.#")
+    var count int
+    // There could be duplicates - worth continuing even if the counts are unequal.
```
Comment (danawillow) [inline templates/terraform/unordered_list_customize_diff.erb]: Shouldn't this be a diff though?
```
Old  New
-a   -a
-a   -b
-b
```

Comment (ndmckinley) [reply templates/terraform/unordered_list_customize_diff.erb]: Things marked as unordered lists are all intended to be changed to sets in 2.0.0 - it's better to not have user-visible changes when that happens.  So I think this should not be a diff!  :)

File: templates/terraform/unordered_list_customize_diff.erb (hunk 6 above)
Comment (ndmckinley) [reply templates/terraform/unordered_list_customize_diff.erb]: Agreed, fixed and working its way through.
